### ISSUE TITLE
Upgrade mechanism

### ISSUE THREAD

**Issue by kaikreuzer** [openhab/openhab-distro#299]
We need to have a stable way to upgrade from release to release as well as from snapshot to snapshot.
First this needs to be solved for the distro, then for apt, docker, spk, etc.

My idea for the upgrade would be the following:
- completely replace the runtime folder by the new version
- delete userdata/cache and userdata/tmp folders
- keep conf and the rest of userdata fully in place

With this, I would assume Karaf to reconstruct the OSGi framework with all the latest versions of the bundles and using the new "openhab-base", while using all statically (conf) and dynamically (userdata) created data as before.

**Comment by kaikreuzer** [comment]
I did some analysis here. A first major problem is https://github.com/openhab/openhab-distro/issues/315 - this causes configuration to be lost.

**Comment by kaikreuzer** [comment]
With https://github.com/openhab/openhab-distro/pull/318, the configuration is moved to userdata/etc, so if the upgrade mechanism leaves this untouched, the "lost configuration issue" should be solved with it.

Next up: If add-ons are installed through Paper UI, these are lost if the cache folder is deleted, see https://github.com/openhab/openhab-distro/issues/319.

**Comment by kaikreuzer** [comment]
My nice theory from above does not really work out - by moving the etc folder from runtime to userdata, we now also have some system files in userdata that might need to be updated on an upgrade. Likewise, we might need to offer new versions of certain conf or userdata files that might have been edited by the user and in case we do folder structure refactorings, we might even have to move stuff around.

@theoweiss explained [here](https://github.com/openhab/openhab-snap/issues/2#issuecomment-2652871859) how this is handled in Debian packages. Unfortunately, we require a mechanism that will work across all our packaging types (Debian, Docker, Synology, QNAP, Snap, etc.) and even on Windows 😦 .

Looking for good suggestions how to deal with that!

**Comment by mhilbush** [comment]
> we now also have some system files in userdata that might need to be updated on an upgrade.

One thing I'll mention here, since I stumbled upon this just yesterday.  I was not seeing in PaperUI any bindings that were newly added to the distro.  It seems I got caught in the middle of this and a couple earlier changes.

https://github.com/openhab/openhab-distro/pull/354

I discovered that userdata/config/profile.cfg was still referencing this:
```
repository.mvn:org.openhab.distro/openhab-features/2.0.0-SNAPSHOT/xml/features="mvn:org.openhab.distro/openhab-features/2.0.0-SNAPSHOT/xml/features"
```

instead of this:
```
repository.mvn:org.openhab.distro/openhab-aggregate-xml/2.0.0-SNAPSHOT/xml/features="mvn:org.openhab.distro/openhab-aggregate-xml/2.0.0-SNAPSHOT/xml/features"
```

I also noticed that the userdata/config directory (and contents) are not part of the distro.  How is profile.config (and potentially other files in this directory) supposed to get updated when upgrading to a new distro?

**Comment by kaikreuzer** [comment]
userdata/config is the internal storage of OSGi ConfigurationAdmin, this must not be updated by anyone.
The "source" for these entries is the userdata/etc folder. You'll find the profile.cfg in there. This is one of the "system files" in userdata that I referred to [above](https://github.com/openhab/openhab-distro/issues/299#issuecomment-268253707).

**Comment by BClark09** [comment]
Continuing conversation from https://github.com/openhab/openhab-docs/issues/273

If all files that _should_ be overwritten in /var/lib/openhab2/etc are listed somewhere in the distro, it should be possible for us to create upgrade .bat and .sh files for manual users, and also use this in the deb packaging.

**Comment by xsnrg** [comment]
I thought about doing this with my shell script, but think it may be easier to just specify the ones that should not be touched as it is a much smaller list.  This seems to be only the key based stuff, though I probably have not discovered everything.  My current version of the script has a list in it like you are proposing, but I fear it is already out of date.

**Comment by kaikreuzer** [comment]
I wonder if there are any clever scripts/mechanisms out there, which would even allow us to define files that might have to be moved to a new folder or that might have to be deleted completely. Only being able to specify overwrite=yes/no might not be enough mid-term...

**Comment by xsnrg** [comment]
It is fairly easy to parse the directory tree from openhab/* down the tree looking for certain file names.  If they are found in two locations, remove the older one.  Is that the kind of thing you are thinking @kaikreuzer ?  We would probably want to include etc/* in the name, provided the ones we are talking about are all in etc, of course.

**Comment by kaikreuzer** [comment]
Symlinks have a huge potential to mess up things - I would not use them for the "regular" installations.

Note that we might move files from userdata/etc to runtime/etc in future (if we manage to get Karaf accept it), so that we more clearly separate static app files from the user data. That's why I am asking for also a possibility to move files around, delete stuff etc.

**Comment by whopperg** [comment]
@kaikreuzer : i updated the script from the docs for me a little bit which you could see here:
https://github.com/openhab/openhab-docs/pull/307#issuecomment-278290828
this is just a starting point, but should get things rolling in the right direction...

**Comment by BClark09** [comment]
There's a couple of problems with that script in terms of sysVinit and systemd compatibility, but the rest seems good. See the [debian scripts](https://github.com/BClark09/openhab-linuxpkg/tree/master/resources/deb/control-runtime) to see how it's done there. Essentially, you'd need to run the stop start and wait processes in `prerm` then `postinst`.

When we have an update.bat (which is done but needs testing) and an updated version of your update.sh in the runtime, we could then use an additional file to list, should be easy to extend from there.

**Comment by ThomDietrich** [comment]
Regarding the file list: I pretty much like the idea! This could be either done in a csv format or something more complex (json/xml). A csv (`operation;file[;destination]`) might already provide the needed flexibility and is easier parsable in low level languages like bash and batch. Any concerns?

**Comment by whopperg** [comment]
>  See the debian scripts to see how it's done there. Essentially, you'd need to run the stop wait processes in prerm, continue on to the rest of your script and then run the start script found in postinst.
Thanks! I will update the script - also we need to check which user is running the openhab service to set the correct permissions.

**Comment by whopperg** [comment]
Would it be possible if we dont have any version numbers in the urls or filenames at all? If so, this would be always consistent thru out the lifetime of openHAB2.

**Comment by kaikreuzer** [comment]
> A csv (operation;file[;destination]) might already provide the needed flexibility 

Sounds like a good suggestion (unless there is already some wide-spread format in use for such things?).

> Would it be possible if we dont have any version numbers in the urls or filenames at all?

I'd think that we can avoid versions for most files (we won't be moving around bundle jars). If we have to add a version, we might also want to specifically target that version (i.e. delete file_1.1.txt, while keeping file_1.2.txt in place).

**Comment by ThomDietrich** [comment]
@whopperg as I said in the other thread, please create a Pull Request with your script.

After you did that we can discuss extensions. Out of my head:

* Decouple Backup, Restore and Upgrade
* Use upgrade-operations.csv file (example file name)

@kaikreuzer I do not know of any wide-spread format nor did I find one on the internet. I can't think of any higher complexity (delete file x only if file y diesn't exist and z was created on christmas eve) demanding a more sophisticated format. Here are all examples I can think of:

* `DELETE;relative/path/to/file`
* `DELETE;$OPENHAB_CONF/pseudo-relative/path/to/file`
* `MOVE;filepath_old;filepath_new`
* `MOVE;filepath_old;filepath_new`
* `TOUCH;some_file_or_folder` (probably not needed, will create folders which might be of interest)

General remarks:

* All commands should be executed in the order they are given in the file
* paths are relative and can start with one of the defined $OPENHAB variables
* spaces and tabs are to be ignored

One additional idea: A second parameter could contain either of `ERROR/WARN/INFO/NONE` telling the upgrade script how to react on a problem, e.g. file to move doesn't exist, ...

What do you think?

**Comment by bdleedy** [comment]
I wrote an upgrade script and posted it in documentation [openhab/openhab-docs#365 ](https://github.com/openhab/openhab-docs/issues/365) only thinking that it would be a temporary help to users until the people that know what they're doing (you guys) could create the REAL solution. Well, it looks like mine might be a good start. I wrote it in Powershell and call it from a batch file for two reasons. One, you can right-click Run as admin and elevate the .ps1 (you can't right click on the .ps1 itself and do that), and two, we can temporarily bypass the Powershell execution policy. If needed, we could probably convert the entire thing back to batch but starting recently I've tried to write as much as I can in Powershell as Microsoft says they will be dropping the command prompt soon and many times it's more powerful and/or graceful.

I based it on what the documentation currently instructs [here](http://docs.openhab.org/installation/windows.html#updating-the-openhab-runtime).

- Stop and delete the openHAB service.
- Prompt for the user to select the current install folder,
- Rename that folder
- Prompt for new .zip file
- Unzip to install directory
- Delete files from backup that should not be restored (this could be converted to pull from.csv)
- Move old files into new install
- Prompt if service is to be setup
- If so start the runtime, prompt the user to run the command, then call batch file.

I've noticed that there are other folders left behind that should be removed as well. for instance...\runtime\system\org\openhab\core\org.openhab.core now has 2.0.0 and 2.0.0-snapshot. These folders should be removed then as well. Is my script helpful?

**Comment by kaikreuzer** [comment]
Hi @whopperg, @bdleedy & all,

Sorry, I lost track of this, although it is pretty important.
As the openHAB 2.1 release is planned soon, I wonder if we can get some basic scripts working to get people from 2.0 release to 2.1 release easily. As it seems, we probably can do without all the csv magic for now, so your basic scripts for Linux & Windows should be a pretty good start.

Instead of following up on #426 and openhab/openhab-docs#365 individually, it is probably better to keep decisions here as a shared place.

For manually upgrading a 2.0 installation, I guess we have to offer the scripts as a separate download (as they aren't part of 2.0).

To keep it simple, I would simplify its logic and leave parts like shutting down the service and doing a backup to the user.

So we would mainly need these steps for a start:
- Download new release as zip to a temp folder
- Delete folders runtime, userdata/cache & userdata/tmp
- unzip the content of the release zip without overwriting any existing file

@whopperg I noticed that your update.sh script is very specific to Linux - it does not work on MacOS; i would be cool if we could have a script that works on both OSes, so that we do not need yet another version of the script.

What do you think? Can we get that working in the next days? Would be great!

**Comment by BClark09** [comment]
> So we would mainly need these steps for a start:
>
>  - Download new release as zip to a temp folder
>  -  Delete folders runtime, userdata/cache & userdata/tmp
>  - unzip the content of the release zip without overwriting any existing file

I can't confirm that this would work on MacOS, but it should do. At the very least we would have:

```
#!/bin/sh

if [ -z "$1" ]; then
  echo "Usage: sh update.sh VersionNumber [openHAB Dir]"
  echo "       e.g. ./update.sh 2.0.0"
  echo "            ./update.sh 2.1.0 /opt/openHAB2"
  exit 1;
fi

if [ -z "$2" ]; then
  WorkingDir="."
  DirError="The update script is not in openHAB's root directory."
else
  WorkingDir="$2"
  DirError="The specified directory is not openHAB's root directory."
fi

if [ ! -d "$WorkingDir/userdata" ] || [ ! -d "$WorkingDir/conf" ]; then
  echo "$DirError"
  echo "Either specify a directory or place this update script in openHAB's root folder."
  exit 1;
fi

OHVersion="$1"
DownloadLocation="https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab%2F$OHVersion%2Fopenhab-$OHVersion.zip"
OutputFile="./openhab-$OHVersion.zip"

echo "Downloading openHAB $OHVersion..."
curl -Lsf "$DownloadLocation" -o "$OutputFile" || { 
  echo "Download failed, version $OHVersion does not exist." 
  exit 1
}

echo "Removing openHAB System Files..."
rm -rf "$WorkingDir/runtime"
rm -f "$WorkingDir/userdata/etc/all.policy"
rm -f "$WorkingDir/userdata/etc/branding.properties"
rm -f "$WorkingDir/userdata/etc/branding-ssh.properties"
rm -f "$WorkingDir/userdata/etc/config.properties"
rm -f "$WorkingDir/userdata/etc/custom.properties"
rm -f "$WorkingDir/userdata/etc/distribution.info"
rm -f "$WorkingDir/userdata/etc/jre.properties"
rm -f "$WorkingDir/userdata/etc/profile.cfg"
rm -f "$WorkingDir/userdata/etc/startup.properties"
rm -f "$WorkingDir/userdata/etc/org.apache.karaf*"

echo "Clearing cache..."
rm -rf "$WorkingDir/userdata/cache"
rm -rf "$WorkingDir/userdata/tmp"

echo "Updating openHAB..."
unzip -nq "$OutputFile" -d "$WorkingDir/"

echo "Deleting temporary files..."
rm -f "$OutputFile"
```

If @whopperg would like to extend on this or integrate parts of what he has then all should be fine!

**Comment by ThomDietrich** [comment]
Hey @BClark09, thanks! The script looks pretty nice to me.
What I am missing is the caring about running openHAB instances. The least the update script should do is have a best guess wether or not openHAB is a running process. Ideally the script checks if an openHAB instance is running from inside the WorkingDir, then stops it using multiple methods, then checks again.
@whopperg already had a few lines regarding this in his script, maybe you can collaborate on that part.

Other suggestions:
* Add a check for unzip and curl (`if command -v foo >/dev/null`)
* Add `set +e` to the beginning of the script, maybe even `set +ex`
* Print errors to stderr (`echo "moo" >&2`)

**Comment by BClark09** [comment]
I agree, @whopperg also had some nice checks to maintain file ownership, which might be necessary.

**Comment by ThomDietrich** [comment]
Also sounds good.
It would be wise to sync permission setting between the different scripts to avoid conflicts or errors. I for one have the following settings inside openHABian: [link](https://github.com/openhab/openhabian/blob/master/openhabian-setup.sh#L567)

**Comment by whopperg** [comment]
Hi All,

I guess we should have someone that owns a Mac and a Linux box for testing this script. As i only got one Linux box (Debian 9) this would be difficult to test and maintain.

**Comment by BClark09** [comment]
Hi @whopperg,

As long as the commands being used are typically UNIX or POSIX compatible, then the script should work on both. AFAIK, the commands that stop your script running on all versions would be:

```
rsync
getfacl
[...]
setfacl

```

but can be replaced with:

```
FileOwner=$(ls -ld "$WorkingDir/userdata" | awk '{print $3}')
FileGroup=$(ls -ld "$WorkingDir/userdata" | awk '{print $4}')
[...]
chown -R "$FileOwner:FileGroup" "$WorkingDir/userdata"
```

I've put up a Gist that you may [view and edit to your liking](https://gist.github.com/BClark09/e255b0605a95f4be833797f6161d896e)

I think at the very minimum it can be used, however I think it would be better and even safer to incorporate some of what @whopperg has done. Particularly, I like the use of `rsync` which is fine to use for MacOS X but not all Unix distributions contain it, can an implementation be added using `cp`, `rm` and `zip`?

In summary: 

@whopperg, would you be willing to edit your script, or add to my example? Also including your stop and start service arguments?
@ThomDietrich, would you agree that we can accomplish tasks in this order:
 1. Use a corrected version of @whopperg's script. If not, use my script as a minimum.
 2. Add a method for backup.
 3. Stopping the script safely if one of the important commands fail.
 4. Run through a list of possible stop tasks.
 5. Add more information, through stderr and echo.

**Comment by kaikreuzer** [comment]
I'll be happy to test any result on Mac, just to be sure that it works there as well.

One question I am contemplating about: Where will we put the scripts in the distribution? In the root next to start.sh/bat? In runtime/bin, which already contains all the Karaf scripts? Or shall we introduce a new folder "bin" in the root?

I would tend to put it in runtime/bin - this is the only option that does not require any further complexity of updates (as "runtime" is already fully replaced by an update) and it would make them available to apt installations as well (not relevant for update, but backup/restore might be interesting for those installations as well). Wdyt?

**Comment by BClark09** [comment]
> One question I am contemplating about: Where will we put the scripts in the distribution?

Good point, the example I posted above assumes in the root, but I too think that runtime/bin might be a better place. Although, the script as it is tries to remove that folder, and it wouldn't be able to delete or update itself. I think we'll need to consider that too!

**Comment by bdleedy** [comment]
Just leaving a note that I am in fact working on the Windows PowerShell version of this.

**Comment by bdleedy** [comment]
@kaikreuzer You say delete the whole runtime folder. The upgrade documentation I followed said to just overwrite all the files but deleting it would solve the issue I referred to above about older files remaining. Delete or overwrite, is there a need to keep any extra files in the runtime folder? (Delete is easier!)

**Comment by kaikreuzer** [comment]
>  it wouldn't be update itself without getting curious errors

Good point, so probably the update script should be copied to some temp folder before execution, so that it is able to update itself as well :-)

@bdleedy Definitely delete! As users must not edit anything in that folder and the runtime itself does not write into it, it should be fully replaced and nothing has to be kept.

What I have missed in my above description is the treatment of the `userdata/etc` folder - there indeed a few files must be overwritten, while others have to be kept; @BClark09's [script above](https://github.com/openhab/openhab-distro/issues/299#issuecomment-309775581) considers this correctly already.

**Comment by BClark09** [comment]
@whopperg, @kaikreuzer and @ThomDietrich,

In the interests of time, shall I continue finishing up the update script? I think it would be most useful to have all that @whopperg has written for backup in it's own file, separate to the update script?

**Comment by bdleedy** [comment]
@kaikreuzer Ok. All of the other deletes I have from the docs is as written in the script above.

@bclark09 I also had backup scripting written as well and was wondering the same thing. Also, I'm trying to keep the Windows update script as similar to yours so I'm looking forward to seeing your updates so that I can modify mine.

**Comment by whopperg** [comment]
Hi @BClark09,
Hi all,
im a little busy this weekend, so i won't be able to update the script. So if anyone want to take my script and extend it, feel free and go ahead.

I'll guess i could do this in the following week.

Cheers and to all a wonderful weekend!

**Comment by BClark09** [comment]
Thanks @whopperg, what I've done is taken out the automatic service stopping processes and made your script more general, finally I've added a self-updating feature. Would you be happy to continue making separate backup and restore scripts when you have the time? What you have already written is all that is necessary but it'll need to be changed slightly so that it's not relying on rsync (perhaps to archive to a tar). @kaikreuzer, would you say that it would be fine to have just the update scripts for now?

Here's what I have so far to test this file would be runtime/bin/update:

```sh
#!/bin/sh

echo "                                         "
echo "#########################################"
echo "       openHAB 2.x.x update script       "
echo "#########################################"
echo "                                         "

setup(){
  ## Keep the script general by allowing the user to provide the version number to download.
  if [ -z "$1" ]; then
    echo "Usage: ./update.sh VersionNumber [openHAB Dir]"
    echo "       e.g. ./update.sh 2.0.0"
    echo "            ./update.sh 2.1.0 /opt/openHAB2"
    echo "            ./update.sh 2.2.0-SNAPSHOT"
    exit 0
  fi

  ## Ask to run as root to prevent us from running sudo in this script.
  if [ "$(id -u)" -ne 0 ]; then
    echo "Please run this script as root!" >&2
    exit 1
  fi

  ## Second parameter can be the openHAB path, if not assume the script is in runtime/bin!
  if [ -z "$2" ]; then
    WorkingDir="../.."
    DirError="The update script is not in openHAB's runtime/bin directory."
  else
    WorkingDir="$2"
    DirError="The specified directory is not openHAB's root directory."
  fi

  ## Check two of the standard openHAB folders to make sure we're updating the right thing.
  if [ ! -d "$WorkingDir/userdata" ] || [ ! -d "$WorkingDir/conf" ]; then
    echo "$DirError" >&2
    echo "Either specify a directory or place this update script in and run from openHAB's runtime/bin folder." >&2
    exit 1;
  fi

  ## Check to see if processes are running before updating
  if [ ! -z "$(ps aux  | grep "openhab2.*java" | grep -v grep)" ]; then
    echo "openHAB is running! Please stop the process before updating." >&2
   exit 1
  fi

  ## Download the specified version from Bintray.
  OHVersion="$1"
  if test "${OHVersion#*-SNAPSHOT}" != "$OHVersion"; then
    DownloadLocation="https://openhab.ci.cloudbees.com/job/openHAB-Distribution/lastSuccessfulBuild/artifact/distributions/openhab/target/openhab-$OHVersion.zip"
  else
    DownloadLocation="https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab%2F$OHVersion%2Fopenhab-$OHVersion.zip"
  fi

  ## Set the temporary directories.
  TempDir="/tmp/openhab"
  OutputFile="$TempDir/openhab-$OHVersion.zip"

  ## Store anything in temporary folders
  echo "Making Temporary Directory"
  mkdir -p "$TempDir" || {
    echo "Failed to make temporary directory: $TempDir" >&2
    exit 1
  }
}

## Download the specified version of openHAB and check for an update script.
download(){
  ## Skip this part if the script was called by an older version of itself.
  if [ "$1" != "--skipnew" ]; then
    echo "Downloading openHAB $OHVersion..."
    curl -Lf# "$DownloadLocation" -o "$OutputFile" || {
      echo "Download failed, version $OHVersion does not exist." >&2
      echo "If you believe this to be an error, please check the openHAB website. (www.openhab.org)"
      exit 1
  }
  ## First check if there's a newer version of this update script
  unzip -qp "$OutputFile" runtime/bin/update > "$TempDir/update" 2>/dev/null && {
    echo "Update script in .zip archive found, using that instead."
    chmod a+x "$TempDir/update"
    "$TempDir/update" "$OHVersion" "$(cd "$WorkingDir" && pwd -P)" "--skipnew"; exit 0
  }
  fi
}

########################################################
## UPDATE SCRIPT AND MODIFICATIONS SHOULD START HERE! ##
########################################################

SpecifiedDir="$1"
SpecifiedVersion="$2"
SkipModifier="$3"

##Run the initialisation functions defined above
setup "$SpecifiedDir" "$SpecifiedVersion"
download "$SkipModifier"

## Preserve file ownership of old setup.
FileOwner=$(ls -ld "$WorkingDir/userdata" | awk '{print $3}')
FileGroup=$(ls -ld "$WorkingDir/userdata" | awk '{print $4}')

## Remove only the files that are to be replaced.
echo "Removing openHAB System Files..."
mkdir -p "$TempDir/runtime"
mkdir -p "$TempDir/userdata/etc"
mv "$WorkingDir/runtime"                                "$TempDir/runtime/"
mv "$WorkingDir/userdata/etc/all.policy"                "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/branding.properties"       "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/branding-ssh.properties"   "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/config.properties"         "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/custom.properties"         "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/distribution.info"         "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/jre.properties"            "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/profile.cfg"               "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/startup.properties"        "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/org.apache.karaf"*         "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/org.ops4j.pax.url.mvn.cfg" "$TempDir/userdata/etc/"

## Clearing the cache and tmp folders is necessary for upgrade.
echo "Clearing cache..."
rm -rf "$WorkingDir/userdata/cache"
rm -rf "$WorkingDir/userdata/tmp"

## Unzip the downloaded folder into openHAB's directory WITHOUT replacing any existing files.
echo "Updating openHAB..."
unzip -nq "$OutputFile" -d "$WorkingDir/" || {
  echo "Failed to unzip archive, restoring system files..." >&2
  ## An error has occured so try to restore openHAB to it's previous state.
  cp -a "$TempDir/runtime"        "$WorkingDir/runtime"
  cp -a "$TempDir/userdata/etc/"* "$WorkingDir/userdata/etc/"
  exit 1
}

## Remove the downloaded zip-file.
echo "Deleting temporary files..."
rm -rf "$TempDir"

## Restore file ownership.
chown -R "$FileOwner:$FileGroup" "$WorkingDir"
```

You can run this script without specifying a directory as long as you're in userdata/bin. Otherwise you can specify the root directory after the version:

```
sudo /opt/openhab2/runtime/bin/update 2.1.0 /opt/openhab2/ 
```

The script tries to find it's newer version of itself within the archive and runs that instead. If there's nothing found it continues with running itself.

@kaikreuzer, would this script work on MacOS? Is this what you had in mind?
@ThomDietrich, would you also be able to test this, and check to see if I've missed anything?
@bdleedy, I've commented on each section, but am not very familiar with batch scripts. Would something like the above be possible?

**Comment by kaikreuzer** [comment]
> @kaikreuzer, would you say that it would be fine to have just the update scripts for now?

Yes, updating is currently our main objective.

> @kaikreuzer, would this script work on MacOS? Is this what you had in mind?

I'll test it tonight and provide feedback - from what I see so far, it is exactly what I have in mind, many thanks! 👍

**Comment by kaikreuzer** [comment]
Just tried it, here's the output:
```
bash-3.2$ sudo ./update.sh 2.1.0-SNAPSHOT

#########################################
       openHAB 2.x.x update script
#########################################

Making Temporary Directory
Downloading openHAB 2.1.0-SNAPSHOT...
######################################################################## 100.0%
Removing openHAB System Files...
mv: rename ../../userdata/etc/all.policy to /tmp/openhab/userdata/etc/all.policy: No such file or directory
mv: rename ../../userdata/etc/branding.properties to /tmp/openhab/userdata/etc/branding.properties: No such file or directory
mv: rename ../../userdata/etc/branding-ssh.properties to /tmp/openhab/userdata/etc/branding-ssh.properties: No such file or directory
mv: rename ../../userdata/etc/config.properties to /tmp/openhab/userdata/etc/config.properties: No such file or directory
mv: rename ../../userdata/etc/custom.properties to /tmp/openhab/userdata/etc/custom.properties: No such file or directory
mv: rename ../../userdata/etc/distribution.info to /tmp/openhab/userdata/etc/distribution.info: No such file or directory
mv: rename ../../userdata/etc/jre.properties to /tmp/openhab/userdata/etc/jre.properties: No such file or directory
mv: rename ../../userdata/etc/profile.cfg to /tmp/openhab/userdata/etc/profile.cfg: No such file or directory
mv: rename ../../userdata/etc/startup.properties to /tmp/openhab/userdata/etc/startup.properties: No such file or directory
mv: rename ../../userdata/etc/org.apache.karaf* to /tmp/openhab/userdata/etc/org.apache.karaf*: No such file or directory
mv: rename ../../userdata/etc/org.ops4j.pax.url.mvn.cfg to /tmp/openhab/userdata/etc/org.ops4j.pax.url.mvn.cfg: No such file or directory
Clearing cache...
Updating openHAB...
Deleting temporary files...
Restoring previous file ownership...
```
The issue seems to be that I called it from within `runtime/bin` and this folder has been replaced during the execution. My terminal doesn't know where it is itself anymore:

```
bash-3.2$ pwd
/Applications/openHAB/runtime/bin
bash-3.2$ cd ..
cd: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
```

**Comment by kaikreuzer** [comment]
Running it from the openHAB root folder, it looks much better:

```
bash-3.2$ sudo ./runtime/bin/update.sh 2.1.0-SNAPSHOT /Applications/openHAB
                                         
#########################################
       openHAB 2.x.x update script       
#########################################
                                         
Making Temporary Directory
Downloading openHAB 2.1.0-SNAPSHOT...
######################################################################## 100.0%
Removing openHAB System Files...
Clearing cache...
Updating openHAB...
Deleting temporary files...
Restoring previous file ownership...
```
And all seems to come up nicely afterwards 👍 
So as long as we tell people NOT to go anywhere into the `runtime` folder for executing the script, all is fine!

**Comment by kaikreuzer** [comment]
Question: Shall we then maybe use the current dir as the assumed project root (and validate this by checking for the existence of `runtime`, `conf` and `userdata`) and thus allow to simply call
```
bash-3.2$ sudo ./runtime/bin/update.sh 2.1.0-SNAPSHOT
```
from there?

Question 2: Would it even be possible to allow an

```
bash-3.2$ sudo ./runtime/bin/update.sh
```
which would then read the version from the `openhab-distro` property of `userdata/etc/version.properties`?

**Comment by BClark09** [comment]
Alright, I edited the above script to do just that, tomorrow I can implement some sort of check with version.properties.

**Comment by BClark09** [comment]
Alright, had some spare time so new script here: Redacted the above to save scrolling:

```sh
#!/bin/sh

setup(){
  ## Keep the script general by allowing the user to provide the version number to download.
  if [ "$1" = help ]; then
    echo "Usage: ./runtime/bin/update [VersionNumber] [openHAB Dir]"
    echo ""
    echo "  e.g. ./runtime/bin/update                << Updates to the next version"
    echo "       ./runtime/bin/update 2.0.0          << Updates to a specific version"
    echo "       ./update 2.1.0 /opt/openHAB2        << Updates a specific root folder"
    echo "       ./runtime/bin/update 2.2.0-SNAPSHOT << Updates to latest SNAPSHOT"
    echo ""
    echo "Use this script to change openHAB to another version. Specifying the version allows"
    echo "you to upgrade or downgrade to that version, or to the latest snapshot. Not specifying"
    echo "any parameters will attempt to find the next version for you."
    exit 0
  fi

  ## Ask to run as root to prevent us from running sudo in this script.
  if [ "$(id -u)" -ne 0 ]; then
    echo "Please run this script as root! (e.g. use sudo)" >&2
    exit 1
  fi

  ## Second parameter can be the openHAB path, if not assume the script is called from root!
  if [ -z "$2" ]; then
    WorkingDir="."
    DirError="The script must be called from openHAB's root directory."
  else
    WorkingDir="$2"
    DirError="The specified directory is not openHAB's root directory."
  fi

  ## Check two of the standard openHAB folders to make sure we're updating the right thing.
  if [ ! -d "$WorkingDir/userdata" ] || [ ! -d "$WorkingDir/conf" ]; then
    echo "$DirError" >&2
    echo "Either specify a directory or place this update script in and run from openHAB's runtime/bin folder." >&2
    exit 1
  fi

  ## Check to see if processes are running before updating
  if [ ! -z "$(ps aux  | grep "openhab2.*java" | grep -v grep)" ]; then
    echo "openHAB is running! Please stop the process before updating." >&2
    exit 1
  fi

  CurrentVersion="$(awk '/openhab-distro/{print $3}' "./userdata/etc/version.properties")"

  OHVersion="$1"

  ## If no OHVersion is specified, try incrementing the second point.
  if [ -z "$OHVersion" ]; then
    FirstPart="$(echo "$CurrentVersion" | awk -F'.' '{print $1}')"
    SecondPart="$(echo "$CurrentVersion" | awk -F'.' '{print $2}')"
    ThirdPart="$(echo "$CurrentVersion" | awk -F'.' '{print $3}')"
    if test "${ThirdPart#*-SNAPSHOT}" != "$ThirdPart"; then
      OHVersion="$CurrentVersion"
    else
      OHVersion="$FirstPart.$((SecondPart + 1)).$ThirdPart"
    fi
  fi

  ## Choose bintray for releases, cloudbees for snapshot.
  if test "${OHVersion#*-SNAPSHOT}" != "$OHVersion"; then
    DownloadLocation="https://openhab.ci.cloudbees.com/job/openHAB-Distribution/lastSuccessfulBuild/artifact/distributions/openhab/target/openhab-$OHVersion.zip"
  elif [ "$OHVersion" = "$CurrentVersion" ]; then
    echo "You are already on openHAB $CurrentVersion" >&2
    exit 1
  else
    DownloadLocation="https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab%2F$OHVersion%2Fopenhab-$OHVersion.zip"
  fi

  ## Set the temporary directories.
  TempDir="/tmp/openhab"
  OutputFile="$TempDir/openhab-$OHVersion.zip"

  ## Store anything in temporary folders
  echo "Making Temporary Directory"
  mkdir -p "$TempDir" || {
    echo "Failed to make temporary directory: $TempDir" >&2
    exit 1
  }
}

## Download the specified version of openHAB and check for an update script.
download(){
  ## Skip this part if the script was called by an older version of itself.
  if [ "$1" != "--skipnew" ]; then
    echo "Downloading openHAB $OHVersion..."
    curl -Lf# "$DownloadLocation" -o "$OutputFile" || {
      echo "Download failed, version $OHVersion does not exist." >&2
      echo "If you believe this to be an error, please check the openHAB website. (www.openhab.org)"
      exit 1
  }
  ## First check if there's a newer version of this update script
  unzip -qp "$OutputFile" runtime/bin/update > "$TempDir/update" 2>/dev/null && {
    echo "Update script in .zip archive found, using that instead."
    chmod a+x "$TempDir/update"
    "$TempDir/update" "$OHVersion" "$(cd "$WorkingDir" && pwd -P)" "--skipnew"; exit 0
  }
  fi
}

########################################################
## UPDATE SCRIPT AND MODIFICATIONS SHOULD START HERE! ##
########################################################

echo "                                         "
echo "#########################################"
echo "       openHAB 2.x.x update script       "
echo "#########################################"
echo "                                         "

SpecifiedDir="$1"
SpecifiedVersion="$2"
SkipModifier="$3"

##Run the initialisation functions defined above
setup "$SpecifiedDir" "$SpecifiedVersion"
download "$SkipModifier"

echo "The script will attempt to update openHAB to version $OHVersion"
echo "Is this okay? [y/N]:"
read answer
case $answer in
  [Yy]*)
    ;;
  *)
    echo "Cancelling update..."
    rm -rf "$TempDir"
    exit 0
    ;;
esac

## Preserve file ownership of old setup.
FileOwner=$(ls -ld "$WorkingDir/userdata" | awk '{print $3}')
FileGroup=$(ls -ld "$WorkingDir/userdata" | awk '{print $4}')

## Remove only the files that are to be replaced.
echo "Removing openHAB System Files..."
mkdir -p "$TempDir/runtime"
mkdir -p "$TempDir/userdata/etc"
mv "$WorkingDir/runtime"                                "$TempDir/runtime/"
mv "$WorkingDir/userdata/etc/all.policy"                "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/branding.properties"       "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/branding-ssh.properties"   "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/config.properties"         "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/custom.properties"         "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/version.properties"        "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/distribution.info"         "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/jre.properties"            "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/profile.cfg"               "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/startup.properties"        "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/org.apache.karaf"*         "$TempDir/userdata/etc/"
mv "$WorkingDir/userdata/etc/org.ops4j.pax.url.mvn.cfg" "$TempDir/userdata/etc/"

## Clearing the cache and tmp folders is necessary for upgrade.
echo "Clearing cache..."
rm -rf "$WorkingDir/userdata/cache"
rm -rf "$WorkingDir/userdata/tmp"

## Unzip the downloaded folder into openHAB's directory WITHOUT replacing any existing files.
echo "Updating openHAB..."
unzip -nq "$OutputFile" -d "$WorkingDir/" || {
  echo "Failed to unzip archive, restoring system files..." >&2
  ## An error has occured so try to restore openHAB to it's previous state.
  cp -a "$TempDir/runtime"        "$WorkingDir/runtime"
  cp -a "$TempDir/userdata/etc/"* "$WorkingDir/userdata/etc/"
  exit 1
}

## Remove the downloaded zip-file.
echo "Deleting temporary files..."
rm -rf "$TempDir"

## Restore file ownership.
echo "Restoring previous file ownership ($FileOwner:$FileGroup)"
chown -R "$FileOwner:$FileGroup" "$WorkingDir"

echo ""
echo "SUCCESS: openHAB updated from $CurrentVersion to $OHVersion"
echo ""
```

Is this a little safer to use on MacOS @kaikreuzer? I've also allowed use of:

```
./update help
```

**Comment by kaikreuzer** [comment]
Thanks, will try that out later today!

**Comment by kaikreuzer** [comment]
```
The script must be called from openHAB's root directory.
Either specify a directory or place this update script in and run from openHAB's runtime/bin folder.
```
This isn't correct, is it? If no directory is specified, it should be run from the root folder.
Running it with a directory from within runtime/bin, it seems to work, but get at the end:
```
SUCCESS: openHAB updated from  to 2.1.0-SNAPSHOT
```
and am again in an outdated folder. So in general, I would not recommend anyone to run it from within the runtime folder.

**Comment by kaikreuzer** [comment]
```
./update help
```
Wouldn't it be more Unix-like to use
```
./update --help
```
or
```
./update -h
```
?

**Comment by BClark09** [comment]
Whoops, forgot to edit that part after making the suggested changes the first time. Done that now, and changed help to --help AND -h.

**Comment by kaikreuzer** [comment]
Thanks @BClark09! 
I'd be fine to go with this script now. Would you want to create a PR to directly add it to https://github.com/openhab/openhab-distro/tree/master/distributions/openhab/src/main/resources/bin? For the upgrade, we could then simply provide the link to file on github.

@bdleedy Could you do the same with the update.bat for Windows?

**Comment by bdleedy** [comment]
I'm looking at my Windows processes and I don't have anything that looks like "openhab2.*java". Anyone have any input on this?

**Comment by bdleedy** [comment]
@BClark09  Can you tell me what you're doing with the version when one isn't provided? Specifically with the snapshot or not snapshot.

**Comment by ThomDietrich** [comment]
Hey @kaikreuzer, @BClark09 and others.
Sorry for not being part of the conversation. My weekend was (as expected) full of non-PC activity. This is my first time of looking at this thread again. The script posted above looks great. How would you expect a user to execute it?

1. go to your openhab directory
2. Execute `sh -c "$(curl -fsSL https://raw.githubusercontent.com/.../upgrade.sh)"`

It would actually be great to remove the first step for most users. You could look at 2-3 usual places ([e.g. `/opt/openhab2`](http://docs.openhab.org/installation/linux.html#manual-installation)), then either execute a quick find or prompt the user to enter the path manually. 

@BClark09 I wonder if you should mention that the script is for manual setups only. Users should not use it on an apt or rpm installation.

**Comment by BClark09** [comment]
@bdleedy 

> I'm looking at my Windows processes and I don't have anything that looks like "openhab2.*java". Anyone have any input on this?

This should work on normal command prompt: Let me know how it goes:

```batch
WMIC path win32_process where "caption='java.exe'" get Commandline | find "openhab"
```

>@BClark09 Can you tell me what you're doing with the version when one isn't provided? Specifically with the snapshot or not snapshot.

Certainly:

```sh
[...]
    if test "${ThirdPart#*-SNAPSHOT}" != "$ThirdPart"; then
```
SNAPSHOT is present in the `version.properties` file, so this line finds the word "SNAPSHOT" and returns true. If the user is already on snapshot, then we just want to update them to snapshot again, so we keep the target version to upgrade the same and continue to the download section.

```sh
  if test "${OHVersion#*-SNAPSHOT}" != "$OHVersion"; then
    DownloadLocation="https://openhab.ci.cloudbees.com/job/openHAB-Distribution/lastSuccessfulBuild/artifact/distributions/openhab/target/openhab-$OHVersion.zip"
  elif [ "$OHVersion" = "$CurrentVersion" ]; then
    echo "You are already on openHAB $CurrentVersion" >&2
    exit 1
  else
    DownloadLocation="https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab%2F$OHVersion%2Fopenhab-$OHVersion.zip"
  fi
```

The download location for each line differs if it's a snapshot or a release version. Again, I'm testing to see if the target version contains the word "SNAPSHOT" then select that download location. Otherwise, from this point if the target version and the current version are the same then we don't need to upgrade.

Hope that helps.

**Comment by BClark09** [comment]
Hey @ThomDietrich, thanks, hope you had a good weekend.

> The script posted above looks great. How would you expect a user to execute it?
> 
> go to your openhab directory
> Execute sh -c "$(curl -fsSL https://raw.githubusercontent.com/.../upgrade.sh)"

That seems like a pretty good way! Personally, I'd rather things be run from the openHAB directory, just to make sure that the obtained location is correct.

> @BClark09 I wonder if you should mention that the script is for manual setups only. Users should not use it on an apt or rpm installation.

I think we can do this in the upgrade instructions, the script will exit without doing anything in an automatic installation because it won't find the correct paths together.

**Comment by ThomDietrich** [comment]
1. Okay then. Let's hope that users know what they are doing :smile_cat: 
2. Ah that's right. Thanks!

**Comment by BClark09** [comment]
> Okay then. Let's hope that users know what they are doing 😸

https://github.com/openhab/openhab-distro/pull/471/commits/a5274d599d6b38525e42736c36049da6ebd3b666 Just in case ;)

**Comment by kaikreuzer** [comment]
FTR, we have a new distro build for testing, which now includes the unix update script: https://openhab.ci.cloudbees.com/job/openHAB-Distribution/967/

**Comment by BClark09** [comment]
![image](https://user-images.githubusercontent.com/8983584/27541312-6574d3b6-5a7b-11e7-9b0f-01bc160d80f9.png)

Works for Linux!

**Comment by kaikreuzer** [comment]
Works for Mac!

**Comment by kaikreuzer** [comment]
> Update script in .zip archive found, using that instead.

This is indeed a very clever move 👍

**Comment by ThomDietrich** [comment]
Excellent! @BClark09 Because I saw the newline in your screenshot just now, this might do better: `read -p "Is this okay? [y/N] " input`

**Comment by BClark09** [comment]
Thanks both.

>> read -p 

`read -p` in POSIX sh is undefined, so didn't use it in case there were some older UNIX machines wanting to update. I also couldn't use `echo -n "" ; read input` for the same reason. Do you think I should change it in the hopes that everyone is using non-archaic systems?

**Comment by ThomDietrich** [comment]
Yeah you are right. I would have ignored that fact because "nobody" is using sh. However it's of course a valid reason. Totally up to you.

**Comment by kaikreuzer** [comment]
@BClark09 One missing feature that comes to my mind: If people have put the openhab-addons-2.0.0.kar or openhab-addons-legacy-2.0.0.kar in their addons folder, this would currently not be replaced - which is an issue as soon as the runtime is started. Would it make sense to check for their existence and replace them with the latest version as well?

**Comment by BClark09** [comment]
Nice find @ThomDietrich! I can use printf, thanks.

> Would it make sense to check for their existence and replace them with the latest version as well?

On it!

**Comment by BClark09** [comment]
@bdleedy, apologies for having to change it a few times already. The latest version of my script can be found [here](https://github.com/openhab/openhab-distro/blob/master/distributions/openhab/src/main/resources/bin/update), which includes the addons file update.

**Comment by ThomDietrich** [comment]
@BClark09 I just realized that the final version of the script is called `update`.
How do you feel about (1) adding `.sh` to clearly mark it as a Linux script file (in parallel to the bat file by @bdleedy) and (2) renaming the script to `upgrade.sh` which is at least in my understanding the more commonly used word for switching the installed version of a software...

**Comment by kaikreuzer** [comment]
1) I thought the same, but then noticed that all other unix scripts in the same folder (most coming from Karaf) do not have .sh either - so I think @BClark09 did it correctly to go with the same convention.
2) Would be ok for me.

**Comment by ThomDietrich** [comment]
1. Ah I see. I should stop making suggestions without a proper assessment of the situation :sweat_smile: Of course @BClark09 already did think of that :1st_place_medal:

**Comment by BClark09** [comment]
I appreciate the suggestions! Always good to check these decisions twice over! :smiley: 

For 2. I chose update because the script can also be used for downgrading (whilst we probably wouldn't recommend it) or switching to snapshot. The word "update" is more general in this sense. It's also the word of choice for MacOS and RPM based Linux systems. I think upgrade is used for apt and firmware only?

**Comment by ThomDietrich** [comment]
I didn't know about Mac or RPM but there are more places where you'll find Upgrade, e.g. https://doc.owncloud.org/server/9.1/admin_manual/maintenance/upgrade.html

Anyhow, we can agree that it's more of a subjective topic and I support your decision. Let's leave it at update. :+1:

**Comment by kaikreuzer** [comment]
> I hope to have a pull request ready tomorrow.

@bdleedy Is "tomorrow" actually today or tomorrow (CET)? Would be great to have it by tonight CET, so that it can be tested and hopefully merged!

**Comment by bdleedy** [comment]
Hammering at this right now. Hoping in about an hour...

**Comment by BClark09** [comment]
Thanks @bdleedy, I'm online and monitoring this issue if you need any input from me.

**Comment by bdleedy** [comment]
Looks like it may be another 24 hours. Sorry, I'll keep you updated.

**Comment by BClark09** [comment]
Is there anything you can give us so far to help with or extend on, it would be great to include this onto the release?

**Comment by ghys** [comment]
I'm available to help with this as well if need be.

**Comment by kaikreuzer** [comment]
@bdleedy Sorry, not possible to wait - we will release today, so I need to do the final build this morning.

Could you share what you have so far so that @ghys could take over?

**Comment by martinvw** [comment]
@BClark09 I just did the upgrade but expected the update script to be in the openHAB directory after the update but it was not there is that expected?

**Comment by kaikreuzer** [comment]
Nobody ever claimed it would work for docker - it is only for "manual" installations!

**Comment by martinvw** [comment]
> Nobody ever claimed it would work for docker - it is only for "manual" installations!

It is my 'production' environment and wanted to test the upgrade with my real sensors so I gave it a try

**Comment by BClark09** [comment]
Sorry, I don't know enough about docker to understand why this didn't work. It looks like it replaced the files correctly, but wouldn't know what those `.dist` directories are, anything out of the expected like this wouldn't work.

**Comment by martinvw** [comment]
> Sorry, I don't know enough about docker to understand why this didn't work. It looks like it replaced the files correctly, but wouldn't know what those .dist directories are, anything out of the expected like this wouldn't work.

Nevermind most likely it is related to docker

**Comment by ghys** [comment]
@bdleedy about the PowerShell update script: due to time constraints - we needed something today - I quickly hacked together another one by taking some bits of @BClark09's [original script](https://github.com/openhab/openhab-docs/issues/365) and following (part of) the logic in the [shell script](https://github.com/openhab/openhab-distro/blob/master/distributions/openhab/src/main/resources/bin/update).

The result was merged earlier today in #473.

Just know my intention wasn't at all to thwart your effort; in fact, there are a few missing pieces in this script, so if what you have already worked on has more features or do things better, I would glady collaborate with you to come up with a final, complete script. 🤝

**Comment by bdleedy** [comment]
@ghys No worries. It looks very similar to my script in many ways. I have a few subtle differences and some larger ones. I was trying to use ExtractToFile but couldn't get that to create folders while not overwriting. I was going to go a route more like yours. I don't see how yours doesn't overwrite though. I also kept as many of my variables and Write-Host outputs the same as the other script for ease of editing later. I'll try to put the two together and drop a message with it in the PR.

**Comment by zarusz** [comment]
I am trying to migrate from OpenHAB 2.0.0-1 to 2.1 on an raspbian according to [this instruction](http://docs.openhab.org/installation/linux.html#upgrade). Invoking the upgrade command:

`sudo sh -c "$(curl -fsSL https://raw.githubusercontent.com/openhab/openhab-distro/master/distributions/openhab/src/main/resources/bin/update)" -- 2.1.0`

Yelds the following error:
```
The script must be called from openHAB's root directory.
Either specify a directory or place this update script in and run from openHAB's root folder.
```

Looking at the script, it does not seem to account for my particular installation layout, which is:
`/etc/openhab2`
`/var/lib/openhab2`
`/usr/share/openhab2`

The script was executed from within `/usr/share/openhab2` which I guess is the root in that case.

Am I doing something wrong or is the script not meant for 2.0 ? Any thoughts?

**Comment by whopperg** [comment]
@zarusz
Hi,

Sure that you got the manual installation of openHAB? 

Looks like you installed it via apt-get or a deb file, I  which case this doesn't work.

If you want to use this, the Installation has to be manual( ie: there is only one path under /opt/openhab2)! 

Cheers

**Comment by BClark09** [comment]
Hi @zarusz, the script is for any **manual** version. It looks like you installed openHAB using apt-get.

If you're still on you should simply be able to update it as if it was any other software on Raspbian using

```
apt-get update && apt-get upgrade
```

If this does not work, then you probably have removed the repository information at some point, just enter in these commands:

```
wget -qO - 'https://bintray.com/user/downloadSubjectPublicKey?username=openhab' | sudo apt-key add -
sudo apt-get install apt-transport-https
echo 'deb https://dl.bintray.com/openhab/apt-repo2 stable main' | sudo tee /etc/apt/sources.list.d/openhab2.list
apt-get update && apt-get upgrade
```

**Comment by whopperg** [comment]
@bclark09 and @all,

Thanks for taking over! I'm just back from Holidays. 
Looks really good to me! 

Do I have to do something or are we good here?

**Comment by zarusz** [comment]
Hi @BClark09  and @whopperg,

Thanks for you prompt response ;)

Yes, indeed I do recall installing it from `apt-get` in the first place (sorry my bad). Yet, somehow `apt-get update && apt-get upgrade` did not work for me so I figured maybe the script will help. However, I just followed @BClark09 script and it works now!

Also I just had to add `sudo` in the last line for it to work:

```
wget -qO - 'https://bintray.com/user/downloadSubjectPublicKey?username=openhab' | sudo apt-key add -
sudo apt-get install apt-transport-https
echo 'deb https://dl.bintray.com/openhab/apt-repo2 stable main' | sudo tee /etc/apt/sources.list.d/openhab2.list
sudo apt-get update && sudo apt-get upgrade
```

Thanks again!

**Comment by BClark09** [comment]
Glad you got it working @zarusz!

@whopperg, there's always something to do around here! ;) In particular I think it'd be good to bundle the next versions of openHAB with a backup and restore scripts. Anyone from this thread should correct me if I'm wrong, but I think the list of things to do for this issue at the moment are as follows:

 - [ ] Create a backup script (you've pretty much done most of this already!).
   - [ ] This should be able to work regardless of installation type (even apt-get!).
 - [ ] Create a restore script.
   - [ ] This should be able to work regardless of installation type (even apt-get!).
 - [ ] Update the windows update script to the same level of functionality as the Linux script.
    - [ ] Allow update of addons files.
    - [ ] Possibly make the script compatible with all versions of powershell or even cmd prompt.
 - [ ] Finalise the idea a csv system to prepare for the update of Karaf

**Comment by BClark09** [comment]
Hi all, will it make sense to make an issue for each of the 4 points above?

**Comment by kaikreuzer** [comment]
Good idea!

**Comment by kaikreuzer** [comment]
Further discussion on the three issues linked above.
